dynamic_programming/fibonacci_sequence_recursion_with_memoization.py

```diff
@@ -0,0 +1,49 @@
+# Fibonacci Sequence Using Recursion and Memoization
+
+# Memoization Concept: https://en.wikipedia.org/wiki/Memoization
+# Fibonacci Concept : https://pt.wikipedia.org/wiki/Sequ%C3%AAncia_de_Fibonacci
+
+
+def fibonacci(n: int, cache: dict = dict()) -> int:
+    """
+    Return number of index n in sequence fibonacci
+    >>> [fibonacci(i) for i in range(12)]
+    [0, 1, 1, 2, 3, 5, 8, 13, 21, 34, 55, 89]
+    >>> fibonacci(12)
+    144
+    """
+    aux = n
+
+    try:
+        if aux in cache:
+            return cache[aux]
+
+        if n in {0, 1}:
+            ans = n
+        else:
+            ans = fibonacci(n - 1, cache) + fibonacci(n - 2, cache)
+        cache[aux] = ans
+
+        return ans
+    except TypeError:
+        print("Error: An integer value is expected")
+    except ValueError:
+        print("Error: The value must be greater than or equal to zero")
```

Comment:
In the above scenario, instead of using the `try ... except ...` pattern, please use the pattern like so,
```python
# pseudocode
if "some condition did not satisfy":
    raise "an error with the appropriate message"
```

Reply:
Corrected, thanks for the remarks